translates: lists-and-keys (#26)
reactjs/nl.reactjs.org: content/docs/lists-and-keys.md

---
id: lists-and-keys

> title: Lijsten En Keys

Reviewer: Comparing with english version

```suggestion
title: Lijsten en Keys
```

permalink: docs/lists-and-keys.html
prev: conditional-rendering.html
next: forms.html
---

Laten we allereerst bekijken hoe je lijsten transformeert in JavaScript.

In de code hieronder gebruiken we de [`map()` (Engels)](https://developer.mozilla.org/nl/docs/Web/JavaScript/Reference/Global_Objects/Array/map) functie op een array van `numbers` waarmee hun waarde wordt verdubbeld. De nieuwe array die door `map()` wordt teruggegeven wijzen we toe aan de variabele `doubled` waarna deze wordt gelogd:

```javascript{2}
const numbers = [1, 2, 3, 4, 5];
const doubled = numbers.map((number) => number * 2);
console.log(doubled);
```

Deze code logt `[2, 4, 6, 8, 10]` naar de console.

In React is het transformeren van arrays naar lijsten van [elementen](/docs/rendering-elements.html) bijna identiek aan het voorgaande.

### Meerdere Componenten Renderen {#rendering-multiple-components}

> Je kunt collecties van elementen opbouwen en [in JSX gebruiken](/docs/introducing-jsx.html#embedding-expressions-in-jsx) met behulp van accolades `{}`.

Reviewer: ```suggestion
Je kunt collecties van elementen maken en [in JSX opnemen](/docs/introducing-jsx.html#embedding-expressions-in-jsx) met behulp van accolades `{}`.
```

---
id: lists-and-keys

> title: Lijsten En Keys

Reviewer 2: ```suggestion
title: Lijsten en Keys
```

Dutch often keeps words like `en` `of` `van` lower-case just like in English, when using `Title Case`

Reviewer 3: We've decided against this style after some discussion in #5, on slack and in various pull requests, see the guidelines in issue #5.

Obviously this discussion is still open, so I'd encourage you to post your thoughts under the issue.

Reviewer 2: Thanks! I read up on it and yep will comment over there :)

permalink: docs/lists-and-keys.html
prev: conditional-rendering.html
next: forms.html
---

> Laten we beginnen met te bekijken hoe je lijsten transformeert in JavaScript.

Reviewer 2: ```suggestion
Laten we beginnen met bekijken hoe je lijsten transformeert in JavaScript.
```

`te` is superfluous here

Reviewer 2: ```suggestion
Laten we allereerst bekijken hoe je lijsten transformeert in JavaScript.
```

Alternatively, this is probably a better translation of the "First/Firstly" idiom.

Reviewer 3: I'm in favor of the last suggestion (with "allereerst") here.

> In de code hieronder gebruiken we de [`map()` (Engels)](https://developer.mozilla.org/nl/docs/Web/JavaScript/Reference/Global_Objects/Array/map) functie op een array van `numbers` waarmee hun waarde wordt verdubbeld.

Reviewer 2: `waarmee` refers back to `een array` which makes this sentence ungrammatical. 

```suggestion
In de code hieronder gebruiken we de [`map()` (Engels)](https://developer.mozilla.org/nl/docs/Web/JavaScript/Reference/Global_Objects/Array/map) functie om de waardes in een array van `numbers` te verdubbelen.
```

> In de code hieronder gebruiken we de [`map()` (Engels)](https://developer.mozilla.org/nl/docs/Web/JavaScript/Reference/Global_Objects/Array/map) functie op een array van `numbers` waarmee hun waarde wordt verdubbeld.

Reviewer: ```suggestion
In de code hieronder gebruiken we de [`map()` (Engels)](https://developer.mozilla.org/nl/docs/Web/JavaScript/Reference/Global_Objects/Array/map) functie om de waarden van een array met `numbers` te verdubbelen.
```

> We wijzen de array die door `map()` wordt teruggegeven toe aan de variabele `doubled` en loggen deze:

Reviewer 2: ```suggestion
We wijzen de nieuwe array die door `map()` wordt teruggegeven toe aan de variabele `doubled` en loggen deze:
```

Perhaps `loggen deze` should be written as `registreren deze` or `leggen deze vast`, but I don't think they're _great_ synonyms.

Reviewer 2: Easier to read might be to change this to 
```suggestion
De nieuwe array die door `map()` wordt teruggegeven wijzen we toe aan de variabele `doubled` waarna deze wordt gelogd:
```

```javascript{2}
const numbers = [1, 2, 3, 4, 5];
const doubled = numbers.map((number) => number * 2);
console.log(doubled);
```

> Deze code logt `[2, 4, 6, 8, 10]` in de console.

Reviewer 2: ```suggestion
Deze code toont `[2, 4, 6, 8, 10]` in de console.
```

Or

```suggestion
Deze code logt `[2, 4, 6, 8, 10]` naar de console.
```

> Het transformeren van arrays naar lijsten van [elementen](/docs/rendering-elements.html) in React is bijna identiek.

Reviewer 2: ```suggestion
In React is het transformeren van arrays naar lijsten van [elementen](/docs/rendering-elements.html) bijna identiek.
```

The question we would also ask ourselves is "identiek waaraan?" so you might want to consider repeating "aan het voorgaande" or something similar: 

```suggestion
In React is het transformeren van arrays naar lijsten van [elementen](/docs/rendering-elements.html) bijna identiek aan het voorgaande.
```

### Meerdere Componenten Renderen {#rendering-multiple-components}

> Je kunt collecties van elementen opbouwen en [in JSX gebruiken](/docs/introducing-jsx.html#embedding-expressions-in-jsx) met behulp van accolades `{}`.

Reviewer 2: I would suggest changing the word order here (which I think would have been better in the original as well):

```suggestion
Met behulp van accolades `{}` kan je collecties van elementen opbouwen en [in JSX gebruiken](/docs/introducing-jsx.html#embedding-expressions-in-jsx).
```

I also think that `opbouwen` isn't a _great_ translation of `build`. I'd almost opt for the less specific `maken` (to make)

> In het voorbeeld hieronder doorlopen we de `numbers` array met de JavaScript [`map()`](https://developer.mozilla.org/nl/docs/Web/JavaScript/Reference/Global_Objects/Array/map) functie.

Reviewer 2: ```suggestion
Hieronder doorlopen we de array van `numbers` gebruik makende van de JavaScript [`map()` (Engels)](https://developer.mozilla.org/nl/docs/Web/JavaScript/Reference/Global_Objects/Array/map) functie.
```

);
```

> We gebruiken de hele `listItems` array in een `<ul>` element en [renderen het naar het DOM](/docs/rendering-elements.html#rendering-an-element-into-the-dom):

Reviewer: ```suggestion
We nemen de hele `listItems` array op in een `<ul>` element en [renderen het naar het DOM](/docs/rendering-elements.html#rendering-an-element-into-the-dom):
```

);
```

[**Probeer het op CodePen**](https://codepen.io/gaearon/pen/GjPyQr?editors=0011)

> Deze code toont een lijst met nummers van 1 tot en met 5.

Reviewer 2: ```suggestion
Deze code toont een lijst met opsommingstekens met nummers van 1 tot en met 5.
```

### Eenvoudige Lijst Component {#basic-list-component}

> Gewoonlijk render je lijsten in een [component](/docs/components-and-props.html).

Reviewer 2: ```suggestion
Normaliter render je lijsten in een [component](/docs/components-and-props.html).
```

Other synonyms: `gangbaar`, `doorgaans` or `in de regel`. I wouldn't use `gewoonlijk` as it has added connotation of being not-special, simple, especially in some dialects.

Or with different word order

```suggestion
Het is gebruikelijk om je lijsten in een [component](/docs/components-and-props.html) te renderen.
```

> We kunnen het eerdere voorbeeld ombouwen naar een component die een array van `numbers` ontvangt en een lijst van elementen weergeeft.

Reviewer 2: ```suggestion
We kunnen het vorige voorbeeld omschrijven naar een component die een array van `numbers` aanneemt en een lijst van elementen teruggeeft.
```

`ontvangt` is probably not correct. The component doesn't `receive` but `accepts to receive / may receive`. I don't like `accepteert`, so perhaps `aanneemt`?

The component doesn't show (`weergeeft`), it `renders`. Either use `weergeven` everywhere where it says `render`, although that is not technically a correct translation, or use `render` or `teruggeeft (outputs)`

);
```

Als je deze code uitvoert, krijg je een waarschuwing dat een "key" nodig is voor lijst elementen. Een "key" is een speciaal text attribuut dat je moet toevoegen als je lijsten van elementen maakt. In de volgende sectie zullen we uitleggen waarom dat belangrijk is.

> Laten we een `key` toevoegen aan onze lijst items in `numbers.map()` en daarmee het probleem van de missende keys oplossen.

Reviewer: ```suggestion
Laten we een `key` toevoegen aan onze lijst elementen in `numbers.map()` en daarmee het probleem dat deze ontbreken oplossen.
```

);
```

> Als je deze code start, krijg je een waarschuwing dat er een "key" nodig is voor de lijst items.

Reviewer 2: ```suggestion
Als je deze code uitvoert, krijg je een waarschuwing dat een "key" nodig is voor lijst elementen.
```

> Een "key" is een speciaal string attribuut dat je moet toevoegen als je lijsten van elementen maakt.

Reviewer 2: ```suggestion
Een "key" is een speciaal text attribuut dat je moet toevoegen als je lijsten van elementen maakt.
```

The original text should probably have codified `string`, in that case:

```suggestion
Een "key" is een speciaal `string` attribuut dat je moet toevoegen als je lijsten van elementen maakt.
```

In de volgende sectie zullen we uitleggen waarom dat belangrijk is.

> Laten we een `key` toevoegen aan onze lijst items in `numbers.map()` en daarmee het probleem van de missende keys oplossen.

Reviewer 2: ```suggestion
Laten we een `key` toewijzen aan onze lijst elementen in `numbers.map()` en daarmee het probleem van de missende "key" oplossen.
```

);
```

[**Probeer het op CodePen**](https://codepen.io/gaearon/pen/jrXYRR?editors=0011)

## Keys {#keys}

> Keys helpen React om items, die veranderd, toegevoegd of verwijderd zijn, te identificeren.

Reviewer 2: ```suggestion
Keys helpen React om items, die veranderd, toegevoegd, of verwijderd zijn, te identificeren.
```

In Dutch writing it would be common to be extra clear using:

```suggestion
Keys helpen React om items - die veranderd, toegevoegd, of verwijderd zijn - te identificeren.
```

But a more literal translation of the original would work too:

```suggestion
Keys helpen React om te identificeren welke elementen zijn veranderd, zijn toegevoegd of zijn verwijderd.
```

> De elementen in een array moeten keys krijgen om ze een stabiele identiteit te geven:

Reviewer 2: The original says `should` so we probably _should_ add `zouden` / `zou` / `zullen`, otherwise we're translating `must`.

```suggestion
De elementen in een array zullen keys moeten krijgen om ze een stabiele identiteit te geven:
```

);
```

> De beste keuze voor een key is een string die het item op unieke wijze identificeert in de lijst.

Reviewer 2: ```suggestion
De beste keuze voor een key is een `string` die het item op unieke wijze identificeert in de lijst.
```

Probably should have been codified in the original, either translate as `tekst` / `woord of woordgroep` or codify.

The current translation removed the `among its siblings`, which was probably superfluous in the original text (as it's unique in the list, therefore among its siblings)

);
```

> Als je geen stabiele IDs hebt voor gerenderde items, kun je als laatste redmiddel ook de item index als key gebruiken:

Reviewer 2: ```suggestion
Als je geen stabiele IDs hebt voor gerenderde items, kun je als laatste redmiddel ook de positie van het element in de lijst als key gebruiken:
```
I would write this out; but can keep index, which would make more sense in the next phrase

```suggestion
Als je geen stabiele IDs hebt voor gerenderde items, kun je als laatste redmiddel ook de index van het element in de lijst als key gebruiken:
```

);
```

> We raden het af om indices als keys te gebruiken als de volgorde van items kan veranderen. 

Reviewer 2: ```suggestion
We raden het af om indices als keys te gebruiken als de volgorde van elementen kan veranderen. 
```

> Dit heeft een negatieve invloed op de performance en kan problemen met de component state veroorzaken.

Reviewer 2: `This` refers to `de volgorde van elementen die veranderd` and not the combination of using indices as keys with that fact -- also incorrect in the original text. We can fix it in the translation by repeating "what" `Dit` refers to:

 ```suggestion
, omdat dit een negatieve invloed heeft op de performance en problemen met de component state kan veroorzaken.
```

When the sentence becomes a composed sentence using the comma, `dit` refers to the _entire first sentence_ which is exactly what we want.

Lees Robin Pokornys artikel voor een [diepgaande uitleg van de negatieve effecten van het gebruik van een index als key (Engels)](https://medium.com/@robinpokorny/index-as-a-key-is-an-anti-pattern-e0349aece318).

> Als je er voor kiest om geen expliciete key aan lijst items te geven, zal React standaard indices gebruiken als keys.

Reviewer 2: ```suggestion
Als je er voor kiest om geen expliciete key aan lijst elementen te geven gebruikt React standaard de indices als keys.
```

`the` is probably missing in the original text

Als je er meer over wilt leren, vind je hier een [diepgaande uitleg waarom keys nodig zijn](/docs/reconciliation.html#recursing-on-children).

> ### Componenten Met Keys Opsplitsen {#extracting-components-with-keys}

Reviewer 2: ```suggestion
### Componenten met Keys Opsplitsen {#extracting-components-with-keys}
```

Keep `met` small.

I like this translation, but more correct would be `extraheren`

```suggestion
### Componenten met Keys Extraheren {#extracting-components-with-keys}
```

**Edit**: that's actually what's used below

Keys zijn alleen logisch in de context van de omliggende array.

Als je bijvoorbeeld een `ListItem` component wilt [extraheren](/docs/components-and-props.html#extracting-components), moet je de key op de `<ListItem />` elementen in de array houden en niet op het `<li>` element in het `ListItem` zelf.

> **Voorbeeld: Foutief Gebruik Van Keys**

Reviewer 2: ```suggestion
**Voorbeeld: Foutief Gebruik van Keys**
```

);
```

> **Voorbeeld: Correct Gebruik Van Keys**

Reviewer 2: ```suggestion
**Voorbeeld: Correct Gebruik van Keys**
```

);
```

[**Probeer het op CodePen**](https://codepen.io/gaearon/pen/ZXeOGM?editors=0010)

Een goede vuistregel is dat de elementen binnen een `map()` aanroep keys nodig hebben.

> ### Keys Moeten Uniek Zijn Ten Opzichte Van Siblings {#keys-must-only-be-unique-among-siblings}

Reviewer 2: ```suggestion
### Keys Moeten Uniek Zijn Ten Opzichte van Naasten {#keys-must-only-be-unique-among-siblings}
```

We don't have a direct translation for `siblings`, but in this case `naasten` might work.

> ### Keys Must Only Be Unique Among Siblings {#keys-must-only-be-unique-among-siblings}

Reviewer 2: ```suggestion
```

Reviewer 4: Keys moeten alleen uniek zijn tussen kinderen (children) van dezelfde ouder (parent)

> Keys die in arrays gebruikt worden moeten uniek zijn ten op zichte van hun siblings.

Reviewer 2: ```suggestion
Keys die in arrays gebruikt worden moeten uniek zijn ten op zichte van hun naasten.
```

);
```

[**Probeer het op CodePen**](https://codepen.io/gaearon/pen/ZXeOGM?editors=0010)

> Een goede vuistregel is dat de elementen binnen een `map()` aanroep keys nodig hebben.

Reviewer 4: Als ik een suggestie mag doen: 
Een vuistregel is dat de elementen binnen de map() functie keys vereisen.
functie is nog niet helemaal goed vind ik zelf, maar operator is dan weer Engels.

);
```

[**Probeer het op CodePen**](https://codepen.io/gaearon/pen/NRZYGN?editors=0010)

> Keys geven een hint aan React maar ze worden niet doorgegeven aan je componenten.

Reviewer 2: ```suggestion
Keys dienen als hint aan React maar ze worden niet doorgegeven aan je componenten.
```

> Als je de waarde nodig hebt in je component, moet je deze expliciet als een prop met een andere naam doorgeven.

Reviewer 2: ```suggestion
Als je dezelfde waarde nodig hebt in je component, moet je deze expliciet als een prop met een andere naam doorgeven.
```

);
```

In het voorbeeld hierboven kan de `Post` component `props.id` lezen, maar `props.key` niet.

> ### Map() Gebruiken In JSX {#embedding-map-in-jsx}

Reviewer 2: ```suggestion
### Map() Gebruiken in JSX {#embedding-map-in-jsx}
```

}
```

> JSX staat toe dat we [elke expressie gebruiken](/docs/introducing-jsx.html#embedding-expressions-in-jsx) tussen accolades, dus  we zouden het `map()` resultaat ook direct kunnen gebruiken:

Reviewer 2: ```suggestion
JSX staat [elke expressie](/docs/introducing-jsx.html#embedding-expressions-in-jsx) toe tussen accolades, dus we zouden het `map()` resultaat ook direct kunnen gebruiken:
```

}
```

[**Probeer het op CodePen**](https://codepen.io/gaearon/pen/BLvYrB?editors=0010)

> Soms levert dit beter leesbare code, maar deze manier kan ook misbruikt worden.

Reviewer 2: ```suggestion
Soms levert dit beter leesbare code op, maar deze stijl kan ook misbruikt worden.
```

Net als in JavaScript is het aan jou om te bepalen of het loont om een variabele te gebruiken voor de leesbaarheid.

> Het kan een goed idee zijn om [een component op te splitsen](/docs/components-and-props.html#extracting-components)  als de `map()` body te diep genest is.

Reviewer 2: ```suggestion
Het kan een goed idee zijn om [een component te extraheren](/docs/components-and-props.html#extracting-components)  als de expressie in `map()` te diep genest is.
```
